src/Api/Vault/AuthorizationHandlers/Collections/BulkCollectionAuthorizationHandler.cs

```diff
@@ -247,19 +257,32 @@ public class BulkCollectionAuthorizationHandler : BulkAuthorizationHandler<BulkC
 
     private async Task<bool> IsAssignedToCollectionsAsync(
         ICollection<Collection> targetCollections,
-        CurrentContextOrganization org,
+        CurrentContextOrganization? org,
         bool requireManagePermission)
     {
         // List of collection Ids the acting user has access to
         var assignedCollectionIds =
             (await _collectionRepository.GetManyByUserIdAsync(_currentContext.UserId!.Value))
             .Where(c =>
                 // Check Collections with Manage permission
-                (!requireManagePermission || c.Manage) && c.OrganizationId == org.Id)
+                (!requireManagePermission || c.Manage) && c.OrganizationId == org?.Id)
             .Select(c => c.Id)
             .ToHashSet();
 
         // Check if the acting user has access to all target collections
         return targetCollections.All(tc => assignedCollectionIds.Contains(tc.Id));
     }
+
+    private async Task<OrganizationAbility?> GetOrganizationAbilityAsync()
+    {
+        if (_organizationAbility != null)
+        {
+            return _organizationAbility;
+        }
+
+        (await _applicationCacheService.GetOrganizationAbilitiesAsync())
+            .TryGetValue(_targetOrganizationId, out var organizationAbility);
+
+        return _organizationAbility = organizationAbility;
+    }
```

Comment:
I'm not sure about the `_organizationAbility` private property here. Presumably it avoids repeated calls to the `applicationCacheService`, but that service is already a cache - it's OK to call it multiple times. (It's just an in-memory dict if you look inside.)

There's also the chance that we check different orgs in the same request (not today afaik but also not impossible). In that case, you'd have the wrong `_organizationAbility` stored here. (That's solvable by comparing orgIds, but as I've said, I don't think it's necessary.)

Reply:
That's fair. Its also the case where we should only be hitting it once per request anyway. 👍